qt-cao-editor/app/scenemodifier.cpp

```diff
@@ -0,0 +1,521 @@
+                                   unsigned int index,float radius, QString load_param)
+{
+    QVector3D main_axis(axis_1[0] - axis_2[0], axis_1[1] - axis_2[1], axis_1[2] - axis_2[2]);
+    QVector3D mid_point((axis_1[0] + axis_2[0])/2, (axis_1[1] + axis_2[1])/2, (axis_1[2] + axis_2[2])/2);
+    QVector3D main_axis_norm = main_axis.normalized();
+
+    Qt3DExtras::QCylinderMesh *cylinder = new Qt3DExtras::QCylinderMesh();
+    cylinder->setRadius(radius);
+    cylinder->setLength(qSqrt(qPow(main_axis[0],2) + qPow(main_axis[1],2) + qPow(main_axis[2],2)));
+    cylinder->setRings(100);
+    cylinder->setSlices(20);
+
+    // CylinderMesh Transform
+    Qt3DCore::QTransform *cylinderTransform = new Qt3DCore::QTransform();
+    cylinderTransform->setRotationX(qAcos(QVector3D::dotProduct(main_axis_norm,QVector3D(1.0f, 0.0f, 0.0f)))*180/M_PI);
+    cylinderTransform->setRotationY(qAcos(QVector3D::dotProduct(main_axis_norm,QVector3D(0.0f, 1.0f, 0.0f)))*180/M_PI);
+    cylinderTransform->setRotationZ(qAcos(QVector3D::dotProduct(main_axis_norm,QVector3D(0.0f, 0.0f, 1.0f)))*180/M_PI);
+    qInfo() << cylinderTransform->rotationX() << cylinderTransform->rotationY() << cylinderTransform->rotationZ() << main_axis[2];
+    //    cylinderTransform->setRotationX(qAcos(main_axis[2]/cylinder->length())*180/M_PI); //theta
+    //    cylinderTransform->setRotationY(qAtan2(main_axis[1],main_axis[0])*180/M_PI);      //phi
+
+    //    cylinderTransform->setRotation(QQuaternion::fromAxisAndAngle(main_axis.normalized(),0.0f));
+    cylinderTransform->setTranslation(mid_point);
+
+    m_cylinderEntity = new Qt3DCore::QEntity(m_rootEntity);
+    m_cylinderEntity->addComponent(cylinder);
+    m_cylinderEntity->addComponent(caoMaterial);
+    m_cylinderEntity->addComponent(cylinderTransform);
+
+    m_cylinderEntity->setObjectName(QString::number(index)+ ":" +load_param);
+    createObjectPickerForEntity(m_cylinderEntity);
+}
+
+void SceneModifier::createCircle(const QVector3D circum_1, const QVector3D circum_2, const QVector3D center,
+                                 unsigned int index, float radius, QString load_param)
+{
+    Qt3DExtras::QTorusMesh *circle = new Qt3DExtras::QTorusMesh();
+    circle->setRadius(radius);
+    circle->setMinorRadius(radius/100);
+    circle->setRings(100);
+    circle->setSlices(20);
+
+    // CircleMesh Transform
+    Qt3DCore::QTransform *circleTransform = new Qt3DCore::QTransform();
+    circleTransform->setTranslation(center);
+
+    m_circleEntity = new Qt3DCore::QEntity(m_rootEntity);
+
+    m_circleEntity->addComponent(circle);
+    m_circleEntity->addComponent(caoMaterial);
+    m_circleEntity->addComponent(circleTransform);
+
+    m_circleEntity->setObjectName(QString::number(index) + ":" + load_param);
+    createObjectPickerForEntity(m_circleEntity);
+}
+
+Qt3DRender::QObjectPicker *SceneModifier::createObjectPickerForEntity(Qt3DCore::QEntity *entity)
+{
+    Qt3DRender::QObjectPicker *picker = nullptr;
```

Comment:
We can use directly `Qt3DRender::QObjectPicker *picker = new Qt3DRender::QObjectPicker(entity);`.